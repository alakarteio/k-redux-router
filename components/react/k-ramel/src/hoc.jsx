import React from 'react'

// TODO: factorize code with react-redux
const getDisplayName = Component => `router(${
  Component.displayName
  || Component.name
  || (Component.constructor && Component.constructor.name)
  || 'Unknown'
})`

const hoc = (code, options) => Component => class extends React.Component {
  static displayName = getDisplayName(Component)

  static contextTypes = {
    store: () => null, // this is to avoid importing prop-types
  }

  constructor(props, context) {
    super(props, context)

    this.state = { show: false }
  }

  componentWillMount() {
    // subscribe
    this.unsubscribe = this.context.store.subscribe(() => {
      this.toShow()
    })

    // run in once
    this.toShow()
  }

  componentWillUnmount() {
    this.unsubscribe()
  }

  toShow = () => {
    const { router } = this.context.store.drivers

    // TODO: hardcoded path (state.ui)
    if (!router.getState() || !router.getResult()) {
      // eslint-disable-next-line no-console
      console.error('[k-redux-router] | There is no route found in `state.ui.router.result`')
      return
    }

    // Absolute mode, we are looking in top level only
    let currentRoute = router.getCurrentRoute()
    if (options && options.absolute) {
      const show = (code === currentRoute.code)
      if (show !== this.state.show) {
        this.setState(innerState => ({ ...innerState, show }))
      }

      return
    }

    // Either way we are looking top down the result tree
    let show = (code === currentRoute.code)
    while (currentRoute && currentRoute.parent && !show) {
      currentRoute = router.getRoute(currentRoute.parent)
      show = (code === currentRoute.code)
    }

    if (show !== this.state.show) {
      this.setState(innerState => ({ ...innerState, show }))
    }
  }

  render() {
    if (!this.state.show) return null

    return <Component {...this.props} />
  }
}

hoc.absolute = (code, options) => hoc(code, { ...options, absolute: true })

export default hoc
