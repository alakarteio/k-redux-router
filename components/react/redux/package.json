{
  "name": "@k-redux-router/react-redux",
  "version": "1.2.0",
  "description": "React components (binded to redux store with k-ramel)",
  "main": "dist/index.umd.js",
  "module": "dist/index.es.js",
  "author": "Fabien JUIF <fabien.juif@gmail.com>",
  "license": "MIT",
  "scripts": {
    "build:core:es": "cross-env NODE_ENV=build rollup -c ./misc/rollup.config.js",
    "build:core:umd": "cross-env NODE_ENV=build FORMAT=umd rollup -c ./misc/rollup.config.js",
    "build": "run-p build:*:*"
  },
  "devDependencies": {
    "@babel/cli": "^7.4.4",
    "@babel/core": "^7.4.5",
    "@babel/plugin-proposal-class-properties": "^7.4.4",
    "@babel/plugin-proposal-object-rest-spread": "^7.4.4",
    "@babel/preset-react": "^7.0.0",
    "cross-env": "^5.1.5",
    "rollup": "^2.10.5",
    "rollup-plugin-babel": "^4.3.2",
    "rollup-plugin-commonjs": "^10.0.0",
    "rollup-plugin-terser": "^5.0.0"
  },
  "peerDependencies": {
    "@k-redux-router/core": "^1.0.0",
    "react-redux": "^5.0.7"
  },
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  },
  "babel": {
    "presets": [
      "@babel/preset-react"
    ],
    "plugins": [
      "@babel/plugin-proposal-object-rest-spread",
      "@babel/plugin-proposal-class-properties"
    ]
  }
}
