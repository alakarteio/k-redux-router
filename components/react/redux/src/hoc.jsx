import React from 'react'
import { selectors } from '@k-redux-router/core'

const getDisplayName = Component => `router(${
  Component.displayName
  || Component.name
  || (Component.constructor && Component.constructor.name)
  || 'Unknown'
})`

const hoc = (code, options = {}) => (Component) => {
  const {
    getState = (state => state.ui.router),
    absolute = false,
  } = options

  const { getResult, getRoute, getCurrentRoute } = selectors(getState)

  return class extends React.Component {
    static displayName = getDisplayName(Component)

    static contextTypes = {
      store: () => null, // this is to avoid importing prop-types
    }

    constructor(props, context) {
      super(props, context)

      this.state = { show: false }
    }

    componentWillMount() {
      // subscribe
      this.unsubscribe = this.context.store.subscribe(() => {
        this.toShow()
      })

      // run in once
      this.toShow()
    }

    componentWillUnmount() {
      this.unsubscribe()
    }

    toShow = () => {
      const state = this.context.store.getState()

      if (!getState(state) || !getResult(state)) {
        // eslint-disable-next-line no-console
        console.error('[k-redux-router] | There is no route found in `state.ui.router.result`')
        return
      }

      // Absolute mode, we are looking in top level only
      let currentRoute = getCurrentRoute(state)
      if (absolute) {
        const show = (code === currentRoute.code)
        if (show !== this.state.show) {
          this.setState(innerState => ({ ...innerState, show }))
        }

        return
      }

      // Either way we are looking top down the result tree
      let show = (code === currentRoute.code)
      while (currentRoute && currentRoute.parent && !show) {
        currentRoute = getRoute(currentRoute.parent)(state)
        show = (code === currentRoute.code)
      }

      if (show !== this.state.show) {
        this.setState(innerState => ({ ...innerState, show }))
      }
    }

    render() {
      if (!this.state.show) return null

      return <Component {...this.props} />
    }
  }
}

hoc.absolute = (code, options) => hoc(code, { ...options, absolute: true })

export default hoc
