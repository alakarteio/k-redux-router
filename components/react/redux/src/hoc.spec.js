/* eslint-disable
  react/jsx-filename-extension,
*/
/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'
import { createStore, combineReducers } from 'redux'
import { Provider } from 'react-redux'
import hoc from './hoc'

const matchApp = (code, multiple = false, notFound) => () => {
  const store = createStore(
    combineReducers({
      ui: combineReducers({
        router: (state = {}) => state,
      }),
    }),
    {
      ui: {
        router: {
          routes: {
            map: {
              relative: {
                code: 'relative',
              },
              absolute: {
                code: 'absolute',
              },
            },
          },
          result: {
            found: !notFound,
            route: {
              code,
              parent: code.split('_')[0], // either relative or absolute or undefined
            },
          },
        },
      },
    },
  )

  const RelativeDecoratedComponent = hoc(multiple ? ['relative', 'dummy'] : 'relative')(() => <div>Relative component</div>)
  const AbsoluteDecoratedComponent = hoc.absolute(multiple ? ['dummy', 'absolute'] : 'absolute')(() => <div>Absolute component</div>)
  const NotFoundDecoratedComponent = hoc.notFound()(() => <div>Not found component</div>)
  const App = () => (
    <Provider store={store}>
      <div>
        <RelativeDecoratedComponent />
        <AbsoluteDecoratedComponent />
        <NotFoundDecoratedComponent />
      </div>
    </Provider>
  )

  expect(renderer.create(<App />)).toMatchSnapshot()
}

describe('hoc', () => {
  describe('uniq code (string)', () => {
    it('should print relative route -strict equality-', matchApp('relative'))
    it('should print relative route -child-', matchApp('relative_child'))
    it('should print absolute route -strict equality-', matchApp('absolute'))
    it('should not print absolute route -child-', matchApp('absolute_child'))
  })

  describe('array of codes', () => {
    it('should print relative route -strict equality-', matchApp('relative', true))
    it('should print relative route -child-', matchApp('relative_child', true))
    it('should print absolute route -strict equality-', matchApp('absolute', true))
    it('should not print absolute route -child-', matchApp('absolute_child', true))
  })

  describe('not found', () => {
    it('should print the failover component when route is not found', matchApp('notFound', false, true))
  })
})
