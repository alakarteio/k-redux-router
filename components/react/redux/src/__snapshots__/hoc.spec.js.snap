// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hoc array of codes should not print absolute route -child- 1`] = `<div />`;

exports[`hoc array of codes should print absolute route -strict equality- 1`] = `
<div>
  <div>
    Absolute component
  </div>
</div>
`;

exports[`hoc array of codes should print relative route -child- 1`] = `
<div>
  <div>
    Relative component
  </div>
</div>
`;

exports[`hoc array of codes should print relative route -strict equality- 1`] = `
<div>
  <div>
    Relative component
  </div>
</div>
`;

exports[`hoc uniq code (string) should not print absolute route -child- 1`] = `<div />`;

exports[`hoc uniq code (string) should print absolute route -strict equality- 1`] = `
<div>
  <div>
    Absolute component
  </div>
</div>
`;

exports[`hoc uniq code (string) should print relative route -child- 1`] = `
<div>
  <div>
    Relative component
  </div>
</div>
`;

exports[`hoc uniq code (string) should print relative route -strict equality- 1`] = `
<div>
  <div>
    Relative component
  </div>
</div>
`;
